concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -0,0 +1,804 @@
+import qualified Concordium.Crypto.VRF as VRF
+import Concordium.Types
+import Concordium.Types.HashableTo
+import Concordium.Types.Transactions
+import Concordium.Utils.Serialization
+
+import qualified Concordium.GlobalState.Basic.BlockState.LFMBTree as LFMBT
+
+-- |A round number for consensus.
+newtype Round = Round {theRound :: Word64}
+    deriving (Eq, Ord, Show, Serialize, Num, Integral, Real, Enum, Bounded)
+
+-- |A strict version of 'Maybe'. We deliberately avoid defining generalised serialization and
+-- hashing instances, so that specific instances can be given as appropriate.
+data Optionally a
+    = Absent
+    | Present !a
+    deriving (Eq, Ord, Show, Functor, Foldable, Traversable)
+
+-- |Returns 'True' if and only if the value is 'Present'.
+isPresent :: Optionally a -> Bool
+isPresent Absent = False
+isPresent (Present _) = True
+
+-- |Returns 'True' if and only if the value is 'Absent'.
+isAbsent :: Optionally a -> Bool
+isAbsent Absent = True
+isAbsent (Present _) = False
+
+-- |The message that is signed by a finalizer to certify a block.
+data QuorumSignatureMessage = QuorumSignatureMessage
+    { -- |Hash of the genesis block.
+      qsmGenesis :: !BlockHash,
+      -- |Hash of the block being signed.
+      qsmBlock :: !BlockHash
+    }
+    deriving (Eq, Show)
+
+-- |Compute the byte representation of a 'QuorumSignatureMessage' that is actually signed.
+-- TODO: check that this cannot collide with other signed messages.
+quorumSignatureMessageBytes :: QuorumSignatureMessage -> BS.ByteString
+quorumSignatureMessageBytes QuorumSignatureMessage{..} = runPut $ do
+    putByteString "QUORUM."
+    put qsmGenesis
+    put qsmBlock
+
+-- |Signature by a finalizer on a 'QuorumSignatureMessage', or an aggregation of signatures on a
+-- common message.
+newtype QuorumSignature = QuorumSignature {theQuorumSignature :: Bls.Signature}
+    deriving (Eq, Ord, Show, Serialize, Semigroup, Monoid)
+
+-- |Sign a 'QuorumSignatureMessage' with a baker's private key.
+signQuorumSignatureMessage :: QuorumSignatureMessage -> BakerAggregationPrivateKey -> QuorumSignature
+signQuorumSignatureMessage msg privKey =
+    QuorumSignature $ Bls.sign (quorumSignatureMessageBytes msg) privKey
+
+-- |Check the signature on a 'QuorumSignatureMessage' that is signed by a single party.
+checkQuorumSignatureSingle ::
+    QuorumSignatureMessage -> BakerAggregationVerifyKey -> QuorumSignature -> Bool
+checkQuorumSignatureSingle msg pubKey =
+    Bls.verify (quorumSignatureMessageBytes msg) pubKey
+        . theQuorumSignature
+
+-- |Check the signature on a 'QuorumSignatureMessage' that is signed by multiple parties.
+checkQuorumSignature ::
+    QuorumSignatureMessage -> [BakerAggregationVerifyKey] -> QuorumSignature -> Bool
+checkQuorumSignature msg pubKeys =
+    Bls.verifyAggregate (quorumSignatureMessageBytes msg) pubKeys
+        . theQuorumSignature
+
+-- |Index of a finalizer in a finalization committee.
```

Comment:
Could you explain what the indexing is, as in, how is the order in the committee defined.